### DISCUSSION TITLE
[3.x] real-time array validation behavior/missing/lacking

### DISCUSSION THREAD

**Discussion by lucasRolff** [livewire/livewire#7839] [7. Bugs (archived)]
### Livewire version

v3.4.1

### Laravel version

v10.42.0

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

- 1: In the wirebox.app link there's two URLs: `/` is using Form Classes and `/working` is using a standard Livewire component with validation

- 2: in `/` if you fill out the 3 fields with value `1`, you'll see no validation trigger on the first 2, the 3rd will give you "The quota field must be between 100 and 1000
- 3: in `/` (FormClass.php), uncomment the `public function updated($field)`, this will cause the initial two fields to get validated, however, the behavior being the previous validation error disappears
- 4: in `/working` (Working.php), fill out the 3 fields with value `1`, we're now seeing errors being kept

### Repository or Wirebox URL

https://wirebox.app/b/4kw6z

### Additional information

In general it seems that `#[Validate]` as a real-time thing does not work with arrays, regardless whether we use form classes or not, even if providing default settings (Suggested in Livewire Discord) such as:
```php
#[Validate]
public array $email_settings = [
    'rspamd_reject_score' => 50,
    'rspamd_subject_score' => 50,
];
```
Then it will not trigger validation for `email_settings.rspamd_reject_score` and `email_settings.rspamd_subject_score` upon updating these fields.

Only if we put the `public function updated($fields)` (in both cases), do we see the array validation triggering. However, when using form classes, it causes unexpected behavior when `public function updated($fields)` is used in the livewire component.

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by calebporzio** [comment]
Are you able to submit a fasting test pr? Or provide further exploration?

**Reply by lucasRolff** [reply]
Hi Caleb,

I'm (sadly) not a very good developer, but if there's anything you possibly want me to check, I can give it a try :)

My hope is that we somehow can get realtime validation to work on arrays again when using form classes, e.g. with the use of `#[Validate]` attributes, but I don't know if it's possible or even sensible to do that, because I'm guessing one would have to find a way to make `#[Validate]` listen for keys in an array as well

In my particular case, I always know what keys in the array may be available, since it's a `settings` array containing a list of pre-defined settings, so in my case I wouldn't mind having to list a default set of "known" keys for the validator to listen for, but I guess other's usecase might be different. In the ideal world I think being fully able to replicate the validateOnly function within form classes as you can do within a component. But I don't know if that's really easily doable:
```php
public function updated($field)
{
    $this->validateOnly($field);
}

**Reply by szajens** [reply]
[https://wirebox.app/b/y82md](https://wirebox.app/b/y82md)

**Reply by joshhanley** [reply]
@szajens can you submit that as a failing test PR? Thanks!

**Reply by szajens** [reply]
Hello, I'm unable to perform this test as I've never done it before. 
Validation works for string, but does not work for a nested string in array. Would you have time to report it?

**Comment by ryanpilbeam** [comment]
Hey. Running into a similar issue with real time validation not working with arrays. I've created an example here; https://wirebox.app/b/4n5k0.

If you submit the form the validation errors show, but as you are switching between inputs no validation errors are shown to the user.

**Reply by pecepl** [reply]
Your Wirebox is private.

**Reply by ryanpilbeam** [reply]
Should be public now.

**Reply by fahrim** [reply]
The same problem applies to me.

**Reply by lrljoe** [reply]
The simple approach - don't use arrays to store inputs!  There's **very rarely** any real use case for this!

Use a "Livewire\Form" class, and use individual properties.  It'll bring you a heap more benefits longer term.

If you **MUST** use an array, then you'll have to just hook into the updated lifecycle hook

e.g.

```
   // Only fire when $billing is updated
    public function updatedBilling($value)
    {
        $this->validate();
    }
```

**Reply by Rikaelus** [reply]
> The simple approach - don't use arrays to store inputs! There's **very rarely** any real use case for this!
> 
> Use a "Livewire\Form" class, and use individual properties. It'll bring you a heap more benefits longer term.
> 
> If you **MUST** use an array, then you'll have to just hook into the updated lifecycle hook
> 
> e.g.
> 
> ```
>    // Only fire when $billing is updated
>     public function updatedBilling($value)
>     {
>         $this->validate();
>     }
> ```

As I've come to learn (so far -- still learning) there's a pretty big use case when dealing with large datasets in child components. Since only one property can be imported into the child via `wire:model`, you pretty much have to use an array (and potentially a nested array) to represent everything the child will deal with.

**Reply by lrljoe** [reply]
The bit I'd question - why are you passing an array of data into a child component?

Can you give a real world example?

**Reply by Rikaelus** [reply]
> The bit I'd question - why are you passing an array of data into a child component?
> 
> Can you give a real world example?

Easy. Look at @ryanpilbeam's [wirebox](https://wirebox.app/b/4n5k0) above and imagine a page that has to collect the billing information of n-number of people in a @foreach loop.

Or alternatively a form a user clicks a + icon on to add n-number of sections dynamically to capture attachments, each section taking a file upload, description, and other metadata.

There are plentiful examples of needing to loop multi-property sections that are ideal candidates for componentization for reusability, whether the components are children and receive Modelable properties or independent and receive properties conventionally.
